libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/RouteCompatibilityCache.kt

```diff
@@ -17,37 +18,51 @@ import com.mapbox.navigation.base.route.NavigationRoute
 object RouteCompatibilityCache {
     private val directionsSessionCache = mutableListOf<NavigationRoute>()
     private val creationCache = LruCache<DirectionsRoute, NavigationRoute>(3)
-    private val lock = Any()
+    private val semaphore = Semaphore(1)
```

Comment:
@cafesilencio, can you elaborate on how this is fixing the issue?

Reply:
I can't recreate the issue but I've seen discussions about synchronized blocks not working well with coroutines. Based on the stack trace something isn't correct with using a generic object to lock on to.  I thought using a Semaphore would be a stronger locking mechanism in this case.  Do you have a suggestion for something different?

Reply:
if we need to sync it with coroutines, we can make the func suspend and use coroutine specific locks:
'ReentrantLock' / `Mutex`

Reply:
> we can't use classes from java.util.concurrent with coroutines

Thanks for the clarification. I knew there was some kind of incompatibility but I couldn't remember the details.

Reply:
> we can't use classes from `java.util.concurrent` with coroutines

There is no such limitation, where did you get that?

Reply:
@korshaknn might know the details better and can comment. I seem to remember that `synchronize` doesn't work with coroutines because they can operate on the same thread (if i'm not mistaken). If you look at the the stack trace in the reported error it seems evident the synchronize blocks aren't working.  There is one call to this class originating in a coroutine.

Reply:
Coroutines provide better alternatives, but synchronization still works

Reply:
I mean` java.util.concurrent` with coroutines will block the whole thread, that's why we should use coroutine specific sync classes
`Mutex.lock()` is a suspending function. It does not block a thread.

Reply:
Doesn't mean we can't use synchronization with coroutines and that it doesn't work